File: nes-coordinator/bridge/coordinator/planner_bridge.rs
/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

use anyhow::{Result, anyhow};
use model::error::{CodedError, ErrorCode};
use model::query::query_fragment::CreateQueryFragment;
use model::statement::Statement;
use sea_orm::DatabaseTransaction;
use tracing::debug;

use crate::TransactionContext;
use crate::ffi;

impl From<ffi::PlannedQueryFragment> for CreateQueryFragment {
    fn from(f: ffi::PlannedQueryFragment) -> Self {
        Self {
            host_addr: f.host_addr.parse().expect("invalid host_addr from planner"),
            plan: f.plan,
            num_operators: f.num_operators,
            has_source: f.has_source,
        }
    }
}

pub(crate) struct FfiSqlPlanner {
    pub(crate) rt_handle: tokio::runtime::Handle,
    pub(crate) optimizer_config: String,
    /// Host substituted for statements that omit their HOST clause.
    /// Empty means such statements are rejected, and a non-empty address is used for embedded deployments.
    pub(crate) default_host: String,
}

impl coordinator::SqlPlanner for FfiSqlPlanner {
    fn plan(
        &self,
        txn: DatabaseTransaction,
        sql: &str,
    ) -> Result<(Statement, DatabaseTransaction)> {
        let ctx = TransactionContext::new(txn, self.rt_handle.clone());

        let planned = ffi::plan_sql(&ctx, sql, &self.optimizer_config, &self.default_host)
            .map_err(|e| anyhow!("{}", e))?;
        if planned.error.code != 0 {
            // The caller is usually a person who wrote the SQL, so they get the message and the code.
            // The C++ stacktrace would be noise to them and goes to the log instead.
            debug!(
                code = planned.error.code,
                trace = %planned.error.trace,
                "planning failed"
            );
            return Err(anyhow::Error::new(CodedError::relayed(
                ErrorCode::from_code(planned.error.code),
                planned.error.msg,
                planned.error.trace,
            )));
        }
        let mut statement: Statement = serde_json::from_str(&planned.json)?;

        if let Statement::CreateQuery(ref mut q) = statement {
            q.fragments = planned
                .fragments
                .into_iter()
                .map(CreateQueryFragment::from)
                .collect();
            q.source_ids = planned.source_ids.into_iter().map(Into::into).collect();
            q.sink_ids = planned.sink_ids.into_iter().map(Into::into).collect();
        }

        Ok((statement, ctx.into_txn()))
    }
}

File: nes-coordinator/bridge/src/PlannerBridge.cpp
/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

#include <PlannerBridge.hpp>

#include <expected>
#include <utility>

#include <Serialization/QueryPlanSerializationUtil.hpp>
#include <nes-coordinator-bridge/catalog.h>
#include <nes-coordinator-bridge/planner.h>
#include <rfl/json/read.hpp>
#include <rfl/json/write.hpp>
#include <rust/cxx.h>
#include <BridgeError.hpp>
#include <CatalogBridge.hpp>
#include <SqlPlanner.hpp>

namespace NES::Bridge
{
namespace
{
std::expected<PlannedStatement, Exception>
planStatement(const PlanningTransaction& ctx, rust::Str sql, rust::Str optimizerConfig, rust::Str defaultHost)
{
    auto sqlStr = std::string(sql.data(), sql.size());
    auto config = QueryOptimizerConfiguration{};
    if (!optimizerConfig.empty())
    {
        auto parsed = rfl::json::read<std::unordered_map<std::string, std::string>>(std::string{optimizerConfig});
        if (!parsed)
        {
            throw CannotDeserialize("Failed to deserialize optimizer config from JSON: {}", parsed.error().what());
        }
        config.overwriteConfigWithCommandLineInput(*parsed);
    }
    /// An empty default host rejects statements that omit their HOST clause.
    /// A non-empty one is substituted (embedded deployments, where the only worker is the local one).
    HostPolicy hostPolicy = RequireHostConfig{};
    if (!defaultHost.empty())
    {
        hostPolicy = DefaultHost{std::string{defaultHost}};
    }
    auto executed = SqlPlanner{std::make_shared<CatalogBridge>(ctx, hostPolicy), std::move(config), std::move(hostPolicy)}.plan(sqlStr);
    if (!executed)
    {
        return std::unexpected{std::move(executed).error()};
    }

    /// The planner's statement is already shaped like the Rust model's, so it serializes without an intermediate conversion.
    auto json = rfl::json::write(executed->statement);

    rust::Vec<PlannedQueryFragment> serialized;
    for (const auto& fragment : executed->fragments)
    {
        const auto bytes = QueryPlanSerializationUtil::serializeQueryPlan(fragment.plan).SerializeAsString();
        rust::Vec<uint8_t> planBytes;
        planBytes.reserve(bytes.size());
        for (const auto byte : bytes)
        {
            planBytes.push_back(static_cast<uint8_t>(byte));
        }
        serialized.push_back(PlannedQueryFragment{
            .host_addr = rust::String(fragment.host.getRawValue()),
            .plan = std::move(planBytes),
            .num_operators = fragment.numOperators,
            .has_source = fragment.hasSource,
        });
    }
    rust::Vec<int64_t> sourceIds;
    for (const auto& id : executed->sourceIds)
    {
        sourceIds.push_back(static_cast<int64_t>(id.getRawValue()));
    }
    rust::Vec<int64_t> sinkIds;
    for (const auto& id : executed->sinkIds)
    {
        sinkIds.push_back(id.getRawValue());
    }

    return PlannedStatement{
        .error = {},
        .json = rust::String(json),
        .fragments = std::move(serialized),
        .source_ids = std::move(sourceIds),
        .sink_ids = std::move(sinkIds),
    };
}
}

PlannedStatement plan_sql(const PlanningTransaction& ctx, rust::Str sql, rust::Str optimizerConfig, rust::Str defaultHost)
{
    return guard([&] { return planStatement(ctx, sql, optimizerConfig, defaultHost); });
}
}
